compiler/rustc_lint/src/array_into_iter.rs

```diff
@@ -70,15 +70,11 @@ impl<'tcx> LateLintPass<'tcx> for ArrayIntoIter {
 
             // Check if the method call actually calls the libcore
             // `IntoIterator::into_iter`.
-            let trait_id = cx
-                .typeck_results()
-                .type_dependent_def_id(expr.hir_id)
-                .and_then(|did| cx.tcx.trait_of_item(did));
-            if trait_id.is_none()
-                || !cx.tcx.is_diagnostic_item(sym::IntoIterator, trait_id.unwrap())
-            {
-                return;
-            }
+            let def_id = cx.typeck_results().type_dependent_def_id(expr.hir_id).unwrap();
```

Comment:
I'm very weary of `.unwrap()` in the compiler. I've had cases where I leaned on it and later changes caused my expectations to not be met.

Reply:
And this is just reverting to what was there before #121586.

I understand the purpose of `span_delayed_bug`, but it shouldn't be used on genuinely unreachable code paths, right? I don't know if you saw #121208 where I deliberately changed a bunch of definitely/probably/possibly unreachable `span_delayed_bug` calls to `span_bug`... some of them were found with subsequent fuzzing to be reachable, which means we got test cases for those paths.